### ISSUE TITLE
Format of Input Data

### ISSUE THREAD

**Issue by wenxiang2001** [ayjxxng/BrainWaveNet#11]
I downloaded the data from the ABIDE preprocessed dataset, which is in the format of multiple 1D files with [time_steps, rois]. How should I convert this data so that it can be directly input into your model? If possible, please provide an example of the input data. Thank you!

**Comment by wenxiang2001** [comment]
I wrote a conversion script to transform the downloaded raw .1D data into a dataset that can be directly read by this program：
```python
import os
import glob
import numpy as np
import pandas as pd
import pickle
import warnings
from sklearn.model_selection import train_test_split
from tqdm.auto import tqdm
import argparse

from cwt_processor import CWT_Processor

warnings.filterwarnings('ignore')

def get_tr_from_site_name(site_name):
    """Get TR value based on site name"""
    TR_MAPPING = {
        'Caltech': 2.0,
        'CMU': 2.0,
        'KKI': 2.5,
        'Leuven': 1.659,
        'MaxMun': 2.0,
        'NYU': 2.0,
        'OHSU': 2.5,
        'Olin': 1.5,
        'Pitt': 1.5,
        'SBL': 2.2,
        'SDSU': 2.0,
        'Stanford': 2.0,
        'Trinity': 2.0,
        'UCLA': 2.0,
        'UM': 2.0,
        'USM': 2.0,
        'Yale': 2.0
    }
    for site in TR_MAPPING:
        if site in site_name:
            return TR_MAPPING[site]
    return 2.0  # Default value

def preprocess_abide_data(
        data_dir, 
        pheno_file, 
        output_dir, 
        test_size=0.2, 
        freq_bands=np.linspace(0.01, 0.1, 5),
        wavelet='cmor1.5-1.0',
        random_state=42
    ):
    """
    Convert ABIDE .1D files to the format required by BrainWaveNet
    
    Parameters:
    - data_dir: Directory containing .1D files
    - pheno_file: Path to the phenotype CSV file
    - output_dir: Output directory
    - test_size: Test set proportion
    - freq_bands: Frequency band settings
    - wavelet: Wavelet type
    - random_state: Random seed
    """
    os.makedirs(output_dir, exist_ok=True)
    
    # 1. Load phenotype data
    print("Loading phenotype data...")
    pheno_df = pd.read_csv(pheno_file)
    
    # 2. Get all .1D files
    file_paths = sorted(glob.glob(os.path.join(data_dir, "*.1D")))
    print(f"Found {len(file_paths)} .1D files")
    
    # 3. Extract information for each file
    all_data = []
    all_labels = []
    all_keys = []  # For stratified sampling
    skipped_files = []

    for file_path in tqdm(file_paths, desc="Processing files"):
        try:
            # Extract subject ID from filename
            filename = os.path.basename(file_path)
            
            # Assuming filename format: site_subjectID_*.1D
            parts = filename.split('_')
            if len(parts) < 2:
                print(f"Unable to parse filename: {filename}")
                skipped_files.append(file_path)
                continue
            
            site = parts[0]
            subject_id = parts[-3]
            
            # Look up diagnostic information in phenotype data
            subject_info = pheno_df[(pheno_df['SUB_ID'] == int(subject_id))]
            
            if len(subject_info) == 0:
                print(f"Subject not found in phenotype data: {site}_{subject_id}")
                skipped_files.append(file_path)
                continue
            
            # Get diagnostic label (1=ASD, 2=Control)
            dx_group = subject_info['DX_GROUP'].values[0]
            # Convert to 0/1 format (0=Control, 1=ASD)
            label = 1 if dx_group == 1 else 0
            
            # Load .1D file data
            data = np.loadtxt(file_path)
            
            # Get TR value for this site
            tr = get_tr_from_site_name(site)
            
            # Save data and label
            all_data.append({
                'data': data,
                'tr': tr,
                'site': site,
                'subject_id': subject_id
            })
            all_labels.append(label)
            all_keys.append(site)  # Use site as key for stratified sampling
            
        except Exception as e:
            print(f"Error processing file {file_path}: {str(e)}")
            skipped_files.append(file_path)
    
    print(f"Successfully processed {len(all_data)} files, skipped {len(skipped_files)} files")
    
    if len(all_data) == 0:
        print("No files were successfully processed, exiting")
        return
    

    
    min_time_points = float('inf')  # Track minimum number of time points
    min_rois = float('inf')  # Track minimum number of ROIs

    # First iterate through data to find minimum time points and minimum ROIs
    for i, item in enumerate(all_data):
        time_points, rois = item['data'].shape
        min_rois = min(min_rois, rois)
        min_time_points = min(min_time_points, time_points)  # Record minimum time points


    # Use the minimum time points found as target length
    target_time_points = min_time_points

    print(f"Number of samples after filtering: {len(all_data)}")
    print(f"Target shape: [time points={target_time_points}, ROIs={min_rois}]")
    
    # 5. Crop data and convert to CWT format
    cwt_data = []
    cwt_labels = []
    cwt_keys = []
    
    for i, item in enumerate(tqdm(all_data, desc="Converting to CWT")):
        # Crop to target shape: take first target_time_points time points
        data = item['data'][:target_time_points, :int(min_rois)]
        tr = item['tr']
        
        # Create temporary dataset dictionary
        temp_dataset = {
            'data': np.array([data]),
            'tr': np.array([tr])
        }
        
        # Instantiate CWT processor
        cwt_processor = CWT_Processor(
            dataset=temp_dataset,
            band=freq_bands,
            wavelet=wavelet
        )
        
        # Get CWT result
        cwt_result = cwt_processor.get_CWT_dataset()
        
        cwt_data.append(cwt_result[0])  # Shape [time_points, freq_bands, ROIs]
        cwt_labels.append(all_labels[i])
        cwt_keys.append(all_keys[i])
    
    # 6. Split into train/valid and test sets
    train_valid_indices, test_indices = train_test_split(
        range(len(cwt_data)), 
        test_size=test_size, 
        random_state=random_state,
        stratify=cwt_labels  # Ensure consistent label distribution
    )
    
    train_valid_data = [cwt_data[i] for i in train_valid_indices]
    train_valid_labs = [cwt_labels[i] for i in train_valid_indices]
    train_valid_keys = [cwt_keys[i] for i in train_valid_indices]
    
    test_data = [cwt_data[i] for i in test_indices]
    test_labs = [cwt_labels[i] for i in test_indices]
    test_keys = [cwt_keys[i] for i in test_indices]
    
    # 7. Save in required format
    print("Saving data...")
    train_valid_path = os.path.join(output_dir, "abide_train_valid_CWT.pkl")
    test_path = os.path.join(output_dir, "abide_test_CWT.pkl")
    
    # Save in format required by the model
    with open(train_valid_path, 'wb') as f:
        pickle.dump((np.array(train_valid_data), np.array(train_valid_labs), train_valid_keys), f)
    
    with open(test_path, 'wb') as f:
        pickle.dump((np.array(test_data), np.array(test_labs), test_keys), f)
    
    print(f"Conversion complete!")
    print(f"Training/validation data: {len(train_valid_data)} samples, saved to {train_valid_path}")
    print(f"Test data: {len(test_data)} samples, saved to {test_path}")
    
    # 8. Save some basic statistics
    stats = {
        'total_samples': len(cwt_data),
        'train_valid_samples': len(train_valid_data),
        'test_samples': len(test_data),
        'labels_distribution': {
            'all': np.bincount(cwt_labels),
            'train_valid': np.bincount(train_valid_labs),
            'test': np.bincount(test_labs)
        },
        'shape': {
            'original': f"[{target_time_points}, {min_rois}]",
            'cwt': f"[{cwt_data[0].shape}]"
        }
    }
    
    with open(os.path.join(output_dir, "preprocessing_stats.pkl"), 'wb') as f:
        pickle.dump(stats, f)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Convert ABIDE .1D files to format usable by BrainWaveNet')
    parser.add_argument('--data_dir', type=str, default="/Volumes/SD10/Neurol/datasets/ABIDE_filt_noglobal", help='Directory containing .1D files')
    parser.add_argument('--pheno_file', type=str, default="/Volumes/SD10/Neurol/datasets/Phenotypic_V1_0b_preprocessed1.csv", help='Path to phenotype CSV file')
    parser.add_argument('--output_dir', type=str, default="/Volumes/SD10/Neurol/datasets/ABIDE_filt_noglobal_output", help='Output directory')
    parser.add_argument('--test_size', type=float, default=0.2, help='Test set proportion')
    parser.add_argument('--random_state', type=int, default=42, help='Random seed')
    
    args = parser.parse_args()
    
    preprocess_abide_data(
        args.data_dir,
        args.pheno_file,
        args.output_dir,
        args.test_size,
        random_state=args.random_state
    )
